### PR TITLE
feat(cloud): cloud task SSE streaming

### PR DESCRIPTION
**Pull request by tatoalo**
## Problem

https://github.com/user-attachments/assets/afa63e15-f088-4cf2-b2c8-8add09ca9f67

Cloud task log viewing used polling (500ms when viewing, 60s if in background) which was inefficient, had dedup complexity, and couldn't stream status changes in realtime. Also it was kinda bad. 
We also had missing comms when a remote sandbox was torn down before the terminal status event arrived, which resulted in ph-code to nevel learn about the run being completed and always show a broken "in progress" state which was not the reflection of the real one.

Added a ton of tests just to make sure this is robust and covers some edge cases.
Diid a lot of manual testing, would appreciate a couple of eyes testing this as well please 🙏🏻 

(https://github.com/PostHog/posthog/pull/53754)

## Changes

- Complete rewrite of `CloudTaskService` from polling to SSE streaming with bootstrap (fetch history + open SSE with `start=latest`), exponential backoff reconnect, and `handleStreamCompletion` that checks actual run status on any stream error
- aadded an sse parse to parse strem chunks with support for `id`, `event`, and multi-line `data` fields
- a bit of refactor to extract `TaskRunStatus`, `TERMINAL_STATUSES`, and `isTerminalStatus()` to `@shared/types` 
- rewrote `watchCloudTask` to pass `apiHost`/`teamId` directly, added `retryCloudTaskWatch`, and `handleCloudTaskUpdate` now handles SSE error events (moved the auth concern to the renderer which already has the auth state, keeping the main-process service simpler)
- sidebar now reads `cloudStatus` from session store (updated via SSE) instead of waiting for API task refetch
- status rename `"started"` → `"queued"` to match BE API

### CODE REVIEW

File: apps/code/src/main/services/cloud-task/service.ts [hunk 1]
```diff
@@ -28,31 +55,123 @@ interface TaskRunResponse {
...
+        {
+          title: "Cloud stream failed",
+          message: `The cloud stream request failed with status ${status}. Retry to reconnect.`,
+          retryable: true,
+          autoRetry: true,
+        },
+        status,
+      );
+  }
+}
```
Comment (jonathanlab) [inline apps/code/src/main/services/cloud-task/service.ts:186]: This is great, love the clear error handling

File: apps/code/src/main/services/cloud-task/service.ts [hunk 2]
```diff
@@ -392,34 +659,291 @@ export class CloudTaskService extends TypedEventEmitter<CloudTaskEvents> {
...
+        taskId: watcher.taskId,
+        runId: watcher.runId,
+        kind: "logs",
+        newEntries: dedupedEntries,
+        totalEntryCount: watcher.totalEntryCount,
+      });
+    }
+
+    watcher.bufferedLogBatches = [];
+  }
```
Comment (jonathanlab) [inline apps/code/src/main/services/cloud-task/service.ts:762]: I wonder if deduplicating by exact content match is the right move or if we should rely purely on the SSE id here..? Not sure, you might have more context than I do.

File: apps/code/src/main/services/cloud-task/sse-parser.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        id: this.currentEventId ?? undefined,
+        data,
+      };
+    } catch {
+      return null;
+    } finally {
+      this.currentEventName = null;
+      this.currentEventId = null;
+    }
+  }
```
Comment (jonathanlab) [inline apps/code/src/main/services/cloud-task/sse-parser.ts:88]: We probably want some log on parse failure here

File: apps/code/src/main/services/cloud-task/service.ts [hunk 4]
```diff
@@ -222,165 +355,299 @@ export class CloudTaskService extends TypedEventEmitter<CloudTaskEvents> {
...
+
+  private handleSseEvent(key: string, event: SseEvent): void {
+    const watcher = this.watchers.get(key);
+    if (!watcher || watcher.failed) return;
+
+    if (event.id) {
+      watcher.lastEventId = event.id;
+    }
+
+    if (event.event === "error") {
```
Comment (jonathanlab) [inline apps/code/src/main/services/cloud-task/service.ts:648]: I see a lot of type casting going on. Could we use a type guard or a Zod schema for this? It makes all this message handling pretty hard to read

File: apps/code/src/main/services/cloud-task/service.ts [hunk 5]
```diff
@@ -222,165 +355,299 @@ export class CloudTaskService extends TypedEventEmitter<CloudTaskEvents> {
...
+      const message =
+        typeof event.data === "object" &&
+        event.data !== null &&
+        "error" in event.data &&
+        typeof (event.data as { error?: unknown }).error === "string"
```
Comment (jonathanlab) [inline apps/code/src/main/services/cloud-task/service.ts]: I see a lot of type casting going on. Could we use a type guard or a Zod schema for this? It makes all this message handling pretty hard to read

File: apps/code/src/renderer/features/sidebar/components/items/TaskItem.tsx [hunk 6]
```diff
@@ -203,6 +199,12 @@ export function TaskItem({
   const isWorktreeTask = workspaceMode === "worktree";
   const isCloudTask = workspaceMode === "cloud";
 
+  const isTerminalCloud =
+    isCloudTask &&
+    (taskRunStatus === "completed" ||
+      taskRunStatus === "failed" ||
+      taskRunStatus === "cancelled");
+
```
Comment (jonathanlab) [inline apps/code/src/renderer/features/sidebar/components/items/TaskItem.tsx]: We can just use the helper you created for this instead

File: apps/code/src/shared/types.ts [hunk 7]
```diff
@@ -104,6 +124,9 @@ export interface CloudTaskUpdatePayload {
   output?: Record<string, unknown> | null;
   errorMessage?: string | null;
   branch?: string | null;
+  // Connection/error fields (present when kind is "error")
+  errorTitle?: string;
+  retryable?: boolean;
 }
```
Comment (jonathanlab) [inline apps/code/src/shared/types.ts]: nit: You could probably turn `CloudTaskUpdatePayload` into a discriminated union to handle all these different cases.

**Review by jonathanlab — APPROVED** [review]
Great stuff. I left mostly nits that would help with keeping the code maintainable/readable.

File: apps/code/src/main/services/cloud-task/service.ts (hunk 2 above)
Comment (tatoalo) [reply apps/code/src/main/services/cloud-task/service.ts:762]: Great point! So in this precise case it is intentional, basically the IDs we work with are redis stream IDs and they are assigned when those events enter the caching system, the historical entries come from S3 and have no stream IDs so the json payload is the only shared key between the sources

Comment (jonathanlab) [reply apps/code/src/main/services/cloud-task/service.ts:762]: ah nice, thanks for explaining
